eip-4626.md

```diff
@@ -0,0 +1,152 @@
+---
+eip: 4626
+title: Yield Bearing Vault Standard
+description: A standard for yield bearing vaults.
+author: Joey Santoro (@joeysantoro), t11s (@transmissions11), Jet Jadeja (@JetJadeja)
+discussions-to: https://ethereum-magicians.org/t/eip-4626-yield-bearing-vault-standard/7900
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-12-22
+---
+
+## Abstract
+
+The following standard allows for the implementation of a standard API for yield bearing vaults within smart contracts. This standard provides basic functionality for depositing and withdrawing tokens and reading balances with an optional extension for tokenized vaults using ERC-20.
+
+## Motivation
+
+Yield bearing vaults have a lack of standardization leading to diverse implementation details. Some various examples include lending markets (Compound, Aave, Fuse), aggregators (Yearn, Rari Vaults, Idle), and intrinsically interest bearing tokens (xSushi). This makes integration difficult at the aggregator or plugin layer for protocols which need to conform to many standards. This forces each protocol to implement their own adapters which are error prone and waste development resources.
+
+A standard for yield bearing vaults will allow for a similar cambrian explosion to ERC-20, unlocking access to yield in a variety of applications with little specialized effort from developers.
+
+
+## Specification
+
+### Methods
+
+#### deposit
+
+`function deposit(address _to, uint256 _value) public returns (uint256 _shares)`
+
+Deposits `_value` tokens into the vault and grants ownership of them to `_to`.
+
+MAY return a pro-rata ownership `_shares` value corresponding `_value`, if not MUST return `0`.
```

Comment:
As it stands now, the inputs to an EIP-4626 pool must be the **underlying**, which is the unit for valuing the pool and the user's position.

I would like to make a case to allow the inputs to be something else, called the **deposit token**. There are a few cases where the **deposit token** could be different from the **underlying**:
1. A vault could take in aUSDC instead of USDC, due to certain design choices. The **underlying** unit of accounting here, is still USDC
2. A wrapper contract could be made to take in gOHM or xSUSHI, use OHM/SUSHI as the underlying unit of accounting, and essentially produce a wrapped_gOhm that adheres to EIP-4626
3. A vault could take in a UniswapV2 LP token, while choosing to use liquidity (sqrt(token1Amount*token2Amount)) as the underlying unit of accounting.

To allow for these use cases, we just need to make one additional view function and three semantic changes:
* Additional function:
     * `depositToken() public view returns (address)`
     * Meaning: the token for depositing and withdrawing, which could be the same as `underlying()`
* Semantic change for depositing:
    * Rename variables from 
       ```
       function deposit(address _to, uint256 _value) public returns (uint256 _shares)
       ```
       to
       ```
       function deposit(address _to, uint256 _amount) public returns (uint256 _shares)
       ```
    * Meaning: deposits `_amount` of deposit tokens, which is worth `_value = _shares * exchangeRate()` in value into the vault and grants ownership of the value to `_to`
* Semantic change for redeeming (shares):
    * Rename variables from 
       ```
       function redeem(address _to, uint256 _shares) public returns (uint256 _value)
       ```
       to
       ```
       function redeem(address _to, uint256 _shares) public returns (uint256 _amount)
       ```
    * Meaning: Redeems a specific number of `_shares` for deposit tokens and transfers them to `_to`. The value of the redeemed tokens will be `_shares * exchangeRate()`
* Semantic change for withdrawing (deposit tokens):
    * Rename 
       ```
       function withdraw(address _to, uint256 _value) public returns (uint256 _shares)
       ```
       to
       ```
       function withdraw(address _to, uint256 _amount) public returns (uint256 _shares)
       ```
    * Meaning: withdraws `_amout` deposit tokens from the vault and transfers them to `_to`.


**In short**: These semantic changes do not alter how things currently work but will make the EIP more general and future-proof.

Reply:
there is no need for this, as what the vault does with the deposited tokens internally is completely irrelevant to this ERC interface. 

underlying signifies what comes in and out of the vault and what totalHoldings() and such is measured in, we do not specify what happens to deposited underlying tokens. you can impl the vaults you proposed without additional metadata.

Reply:
Let's say we want to implement this vault A:
- User deposits cUSDC
- The vault claims COMP over time and convert to more cUSDC

Option 1: If we use the current standard:
- underlying() = cUSDC
- exchangeRate() = rate of how much more cUSDC a user will get, due to COMP
- underlyingBalanceOf(user) = cUSDC balance

Option 2: if we apply the proposed semantic changes here, it can be:
- underlying() = USDC
- exchangeRate() = rate of how much more USDC user is getting, from both Compound's rate and claiming COMP in this vault
- underlyingBalanceOf(user) = USDC-equivalent balance of the user

If we take the perspective of let's say a UI to show the details of this vault, it makes more sense to show the user balance in USDC using underlyingBalanceOf(), and easier to compare this vault with other vaults with the same USDC underlying.
Another example is from the perspective of a protocol to tokenise the yield of this vault with respect to USDC, where Option 2 makes a lot more sense as well.

Reply:
I think that explicitly specifying a multi-layer approach adds complexity to the standard, which should be kept as simple as possible. Not all existing vaults will be immediately compliant and that must be accepted.

As such, I don't think that a `depositToken()` method should be added.

On specifying inputs based on `value` or `shares`, that I agree for the sake of completeness.

1. Deposit exact `value` underlying, receive some `shares`
2. Receive exact `shares`, deposit underlying for some `value`
3. Withdraw exact `shares`, receive underlying for some `value`
4. Receive exact `value` of underlying, by withdrawing some `shares`.

This follows the pattern seen in AMMs that also deal with two tokens (in this case the underlying and the vault tokens).

Externally, the vault does the same job of exchanging between the two, and it is irrelevant the internal mechanism used to determine the exchange rate.

I'm not arguing to provide an interface identical to any existing AMM, but a familiar set of methods, proven to work in an equivalent use case, would definitely help with integration.

Reply:
I think it makes a lot of sense to add a "mint" method which does `3. Withdraw exact shares, receive underlying for some value` @transmissions11 @JetJadeja wdyt?

Reply:
@mrenoon Perhaps whatever is using this interface can go through two interfaces to determine the value in USDC. First determine SHARES/aUSDC value via ERC-4626 and then determine aUSDC/USDC via say a DEX or another ERC-4626. This is kind of composing two "valuation" interfaces which have "calculate" functions. Could that be standardized independent of the rest of EIP-4626? Both ERC-4626 and a DEX could implement this "valuation" interface. Then the logic which composes two valuation interfaces could be wrapped up into its own valuation interface, achieving your goal of going through this one interface which calculates the SHARES/USDC valuation directly.